File: cpp/brackets.cpp
#include <iostream>

using namespace std;

const int MAX = 100;

bool brackets() {
    int sequence[MAX];
    int top = 0;
    int bracket_type;

    cin >> bracket_type;
    while (bracket_type != 0) {

        if (bracket_type > 0)
            sequence[top++] = bracket_type;
        else
            if (bracket_type != -sequence[--top])
                return false;

        cin >> bracket_type;
    }
    if (top != 0)
        return false;
    return true;
}

int main() {
    cout << "enter a bracket sequence\n"
        << "use positive numbers for an open bracket and negative numbers for a closing bracket\n"
        << "0 means end of input\n";

    if (brackets())
        cout << "this is correct bracket sequence\n";
    else
        cout << "this is incorrect bracket sequence\n";

    return 0;
}

File: python/brackets.py
def is_correct_sequence(sequence: str) -> bool:
    stack = []
    for element in sequence:
        if element in '([{':
            stack.append(element)
        else:
            if element in ')]}':
                if not stack:
                    return False

                open_bracket = stack.pop()
                if open_bracket == '(' and element == ')':
                    continue
                if open_bracket == '[' and element == ']':
                    continue
                if open_bracket == '{' and element == '}':
                    continue
                return False

    if not stack:
        return True
    else:
        return False


if __name__ == '__main__':
    data = input()
    if is_correct_sequence(data):
        print('is correct')
    else:
        print('is not correct')